security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -807,9 +808,11 @@ func (sc *SecretCache) generateSecret(ctx context.Context, token string, connKey
 	}, nil
 }
 
-func (sc *SecretCache) shouldRefresh(s *model.SecretItem) bool {
-	// secret should be refreshed before it expired, SecretRefreshGraceDuration is the grace period;
-	return time.Now().After(s.ExpireTime.Add(-sc.configOptions.SecretRefreshGraceDuration))
+func (sc *SecretCache) shouldRotate(secret *model.SecretItem) bool {
+	// secret should be rotated before it expired.
+	secretLifeTime := secret.ExpireTime.Sub(secret.CreatedTime)
+	gracePeriod := secretLifeTime * time.Duration(sc.configOptions.SecretRotationGracePeriodRatio)
+	return time.Now().After(secret.ExpireTime.Add(-gracePeriod))
```

Comment:
Is "-gracePeriod" equivalent to "-1*gracePeriod"? Is secret.ExpireTime.Add(-gracePeriod) equivalent to secret.ExpireTime.Sub(gracePeriod)?

Reply:
Nope. ExpireTime.Sub(gracePeriod) won't work. It doesn't take a Duration as a parameter but Add() does.